src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,262 @@
+            }
+
+            var listObjects = new List<T>();
+            do
+            {
+                listObjects.Add(streamReaderConverter.Read(ref this));
+                //At this point we're looking at the EndObject token for the object, need to advance.
+                Read();
+            }
+            while (TokenType != JsonTokenType.EndObject);
+            return listObjects;
+        }
+
+        internal void ReadArrayOfObjects<T1, T2>(IList<T2> objectList, Utf8JsonStreamReaderConverter<T1> streamReaderConverter) where T1 : T2
+        {
+            if (objectList is null)
+            {
+                return;
+            }
+
+            if (Read() && TokenType == JsonTokenType.StartArray)
+            {
+                while (Read() && TokenType != JsonTokenType.EndArray)
+                {
+                    var convertedObject = streamReaderConverter.Read(ref this);
+                    if (convertedObject != null)
+                    {
+                        objectList.Add(convertedObject);
+                    }
+                }
+            }
+        }
+
+        internal string ReadNextTokenAsString()
+        {
+            if (Read())
+            {
+                return _reader.ReadTokenAsString();
+            }
+
+            return null;
+        }
+
+        internal string GetCurrentBufferAsString() => Encoding.UTF8.GetString(_buffer);
+
+        internal IList<string> ReadStringArrayAsIList(IList<string> strings = null)
+        {
+            if (TokenType == JsonTokenType.StartArray)
+            {
+                while (Read() && TokenType != JsonTokenType.EndArray)
+                {
+                    string value = _reader.ReadTokenAsString();
+
+                    strings = strings ?? new List<string>();
+
+                    strings.Add(value);
+                }
+            }
+
+            return strings;
+        }
+
+        internal IReadOnlyList<string> ReadDelimitedString()
+        {
+            if (Read())
+            {
+                switch (TokenType)
+                {
+                    case JsonTokenType.String:
+                        var value = GetString();
+
+                        return value.Split(DelimitedStringDelimiters, StringSplitOptions.RemoveEmptyEntries);
+
+                    default:
+                        var invalidCastException = new InvalidCastException();
+                        throw new JsonException(invalidCastException.Message, invalidCastException);
+                }
+            }
+
+            return null;
+        }
+
+        internal List<string> ReadStringArrayAsList(List<string> strings = null)
+        {
+            if (TokenType == JsonTokenType.StartArray)
+            {
+                while (Read() && TokenType != JsonTokenType.EndArray)
+                {
+                    string value = _reader.ReadTokenAsString();
+
+                    strings = strings ?? new List<string>();
```

Comment:
I would love better default capacity on every single List<string> creation based on most likely end result to reduce allocations due to resizes and reduce memory usage.

Reply:
@davkean This might be interesting for that: https://github.com/NuGet/Home/pull/12124/files#diff-0e899c815846f62b058281b85f22b98df554f592455dcd732e44c72205ab7cc8R246. 

I'm sure @zivkan would love to know what you think.